working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,948 @@
+the semantics of invocations of the existing extension methods:
+
+If `e` is an expression whose static type is the extension type
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code>,
+then a member access like `e.m(args)` is executed by evaluating `e` to an
+object which is bound to a fresh variable `v`, and then evaluating
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>>(v).m(args)</code>,
+and similarly for instance getters and operators.
+
+The dynamic semantics of an invocation of an instance method of the on-type
+which is enabled in an explicit extension type by the show/hide part is as
+if a forwarder were implicitly induced in the extension, with the same
+signature as that of the on-type. *For example:*
+
+```dart
+// Extension type using show/hide:
+extension type MyNum on num show floor {}
+
+// Works like the following:
+extension type MyNum on num {
+  int floor() => this.floor();
+}
+```
+
+*Note that this implies that the extension method `floor` never overrides
+the instance member `floor`, but the extension method `floor` will be
+executed at a call site `myNum.floor()` based on a compile-time decision
+when the receiver `myNum` has static type `MyNum`. In particular, the
+extension method `floor` will never be executed when the receiver has type
+`dynamic`. The forwarding expression `this.floor()` in the implicitly
+induced getter will invoke the instance method, which is subject to late
+binding (so we may end up running `int.floor()` or `double.floor()`,
+depending on the dynamic type of `this`).*
+
+At run time, for a given instance `o` typed as an extension type `E`, there
+is _no_ reification of `E` associated with `o`.
+
+*This means that, at run time, an object never "knows" that it is being
+viewed as having an extension type. By soundness, the run-time type of `o`
+will be a subtype of the on-type of `E`.*
+
+The run-time representation of a type argument which is a non-protected
+extension type `E` (respectively 
+<code>E<T<sub>1</sub>, .. T<sub>k</sub>></code>) 
+is the corresponding instantiated on-type.
+
+*This means that a non-protected extension type and the underlying on-type
+are considered as being the same type at run time. So we can freely use a
+cast to introduce or discard the extension type, as the static type of an
+instance, or as a type argument in the static type of a data structure or
+function involving the extension type.*
+
+The run-time representation of a type argument which is a protected
+extension type `E` (respectively 
+<code>E<T<sub>1</sub>, .. T<sub>k</sub>></code>) 
+is an identification of `E` (respectively 
+<code>E<T<sub>1</sub>, .. T<sub>k</sub>></code>).
+
+*In particular, it is not the same as the run-time representation of the
+corresponding on-type. This is necessary in order to maintain that the
+on-type and the protected extension type are unrelated.*
+
+*For a protected extension type `E`, with a data structure or function where
```

Comment:
I don't know what to make of this paragraph.  Are you claiming that casting a `List<MyExt>` to `List<Object>` is supposed to fail at runtime?  If so, this is a huge piece of the design that needs **much** more than a paragraph of development.

Reply:
Yes, when `MyExt` is a protected extension type it would be reified as a type distinct from the on-type. So a `List<MyExt>` would not be a `List<Object>`, it would only be a `List<Object?>`.

Reply:
My point was not `Object` vs `Object?`, sorry, I should have been precise.  I was asking about `List<Object?>`.  I'm very confused now because your answer contradicts this paragraph.

Reply:
I'm currently adding further discussion about the design choices here, and also that this kind of run-time failure is implied by some design choices.